src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,503 @@
+	 */
+	private int invocations = 0;
+
+	public ConstantPropogation(Green solver) {
+		super(solver);
+	}
+
+	@Override
+	public Set<Instance> processRequest(Instance instance) {
+		@SuppressWarnings("unchecked")
+		Set<Instance> result = (Set<Instance>) instance.getData(getClass());
+		if (result == null) {
+			final Map<Variable, Variable> map = new HashMap<Variable, Variable>();
+			final Expression e = propagate(instance.getFullExpression(), map);
+			final Instance i = new Instance(getSolver(), instance.getSource(), null, e);
+			result = Collections.singleton(i);
+			instance.setData(getClass(), result);
+		}
+		return result;
+	}
+
+	@Override
+	public void report(Reporter reporter) {
+		reporter.report(getClass().getSimpleName(), "invocations = " + invocations);
+	}
+
+	public Expression propagate(Expression expression,
+			Map<Variable, Variable> map) {
+		try {
+			log.log(Level.FINEST, "Before Propagation: " + expression);
+			invocations++;
+            boolean changed = true;
+            Map<IntVariable, IntConstant> constants = new HashMap<IntVariable, IntConstant>();
+
+            int n = 0;
+            while (changed) {
+                ConstantVisitor constantVisitor = new ConstantVisitor(constants);
+                OrderingVisitor orderingVisitor = new OrderingVisitor();
+                SimplifyVisitor simplifyVisitor = new SimplifyVisitor();
+
+                expression.accept(constantVisitor);
+                expression = constantVisitor.getExpression();
+                expression.accept(orderingVisitor);
+                expression = orderingVisitor.getExpression();
+                expression.accept(simplifyVisitor);
+                expression = simplifyVisitor.getExpression();
+                changed = orderingVisitor.hasChanged() || 
+                    constantVisitor.hasChanged() ||
+                    simplifyVisitor.hasChanged();
+
+                if (n++ > 50) {
+                    log.log(Level.SEVERE, "Failed to simplify further after 50 attempts");
+                    break;
+                }
+
+            }
+
+			log.log(Level.FINEST, "After Propagation: " + expression);
+			return expression;
+		} catch (VisitorException x) {
+			log.log(Level.SEVERE,
+					"encountered an exception -- this should not be happening!",
+					x);
+		}
+		return null;
+	}
+
+	private static class OrderingVisitor extends Visitor {
+
+		private Stack<Expression> stack;
+        private boolean changed;
+
+		public OrderingVisitor() {
+			stack = new Stack<Expression>();
+            changed = false;
+		}
+
+		public boolean hasChanged() {
+			return changed;
+		}
+
+		public Expression getExpression() {
+			return stack.pop();
+		}
+
+		@Override
+		public void postVisit(IntConstant constant) {
+			stack.push(constant);
+		}
+
+		@Override
+		public void postVisit(IntVariable variable) {
+			stack.push(variable);
+		}
+
+		@Override
+		public void postVisit(Operation operation) throws VisitorException {
```

Comment:
Write some Javadoc for this method, it has a quite a bit of meat and a brief explanation of what it's responsibility is would help readability quite a bit.